include/dr/mhp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,275 @@
+
+    rng::advance(s_first, dist - 1);
+
+#ifdef SYCL_LANGUAGE_VERSION
+    if (mhp::use_sycl()) {
+      events.push_back(mhp::sycl_queue().submit([&](auto &&h) {
+        h.depends_on(one_segment_scan_event);
+        h.single_task([=]() {
+          OVal tmp = *s_first;
+          *local_partial_sums_iter = tmp;
+        });
+      }));
+    } else
+#endif
+      *local_partial_sums_iter = *s_first;
+    ++local_partial_sums_iter;
+  }
+
+#ifdef SYCL_LANGUAGE_VERSION
+  sycl::event::wait(events);
+  events.clear(); // events are reused later in this function
+#endif
+
+  const std::size_t local_partial_sums_count =
+      rng::distance(local_partial_sums, local_partial_sums_iter);
+
+  OVal *local_partial_sums_scanned =
+      local_partial_sums_count
+          ? default_allocator<OVal>().allocate(local_partial_sums_count)
+          : nullptr;
+
+  if (local_partial_sums_count) {
+#ifdef SYCL_LANGUAGE_VERSION
+    if (mhp::use_sycl())
+      oneapi::dpl::experimental::inclusive_scan_async(
+          dpl_policy(), local_partial_sums,
+          local_partial_sums + local_partial_sums_count,
+          local_partial_sums_scanned, binary_op)
+          .wait();
+    else
+#endif
+      std::inclusive_scan(std::execution::par_unseq, local_partial_sums,
+                          local_partial_sums + local_partial_sums_count,
+                          local_partial_sums_scanned, binary_op);
+  }
+
+  dr::communicator &comm = default_comm();
+  std::optional<OVal> local_partial_sum =
+      local_partial_sums_count
+          ? local_partial_sums_scanned[local_partial_sums_count - 1]
+          : std::optional<OVal>();
+
+  // below vector is used on 0 rank only but who cares
+  std::vector<std::optional<OVal>> partial_sums(comm.size()); // dr-style ignore
+
+  comm.gather(local_partial_sum, partial_sums, 0);
+
+  if (comm.rank() == 0) {
+    std::optional<OVal> next_v = init;
+    rng::for_each(partial_sums, [&next_v, binary_op](std::optional<OVal> &v) {
+      if (v.has_value()) {
+        std::swap(v, next_v);
+        if (v.has_value()) // mind v was a next line before
+          next_v = binary_op(v.value(), next_v.value());
+      }
+    });
+  }
+
+  // scatter partial_sums_scanned
+  std::optional<OVal> sum_of_all_guys_before_my_rank;
+  comm.scatter(partial_sums, sum_of_all_guys_before_my_rank, 0);
+
+  if (local_partial_sums_count && sum_of_all_guys_before_my_rank.has_value()) {
+    const OVal sum_of_all_guys_before_my_rank_value =
+        sum_of_all_guys_before_my_rank.value();
+#ifdef SYCL_LANGUAGE_VERSION
+    if (mhp::use_sycl())
+      oneapi::dpl::experimental::for_each_async(
+          dpl_policy(), local_partial_sums_scanned,
+          local_partial_sums_scanned + local_partial_sums_count,
+          [=](auto &&x) {
+            x = binary_op(sum_of_all_guys_before_my_rank_value, x);
+          })
+          .wait();
+    else
+#endif
+      std::for_each(std::execution::par_unseq, local_partial_sums_scanned,
```

Comment:
There is an inclusive scan variant that takes an init value. Won't that work here instead of using for_each?

Reply:
No. I want to add constant to each element in this place. Not to compute partial sum. I've explained it in description of alg in comment before inclusive_scan_impl function.

Reply:
I understand now.